File: typescript/15.3-sum.ts
/*
 * @lc app=leetcode id=15 lang=typescript
 *
 * [15] 3Sum
 */

// @lc code=start
function threeSum(nums: number[]): number[][] {
    function twoSum(nums: number[], i: number, results: number[][]) {
        let left = i+1, right = nums.length-1;
        while (left < right) {
            let sum = nums[i] + nums[left] + nums[right];
            if (sum < ) left += 1;
            else if (sum > 0) right -= 1;
            else {
                results.push([nums[i], nums[left], nums[right]]);
                left += 1;
                right -= 1;
                while (nums[left] === nums[left-1]) left += 1;
            }
        }
    }
    nums.sort((a, b) => a-b);
    let results: number[][] = [];
    for (let i = 0; i < nums.length; i += 1) {
        // fix nums[i] and do a two sum
        while (i > 0 && nums[i] === nums[i-1]) i += 1;
        twoSum(nums, i, results);
    }

    return results;
};
// @lc code=end

File: cpp/15.3-sum.cpp
/*
 * @lc app=leetcode id=15 lang=cpp
 *
 * [15] 3Sum
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        for(int i = 0; i < nums.size(); i++) {
            while (i > 0 && i < nums.size()-1 && nums[i] == nums[i-1]) i++;
            twoSum(&nums[i+1], &*(nums.end()-1), -nums[i], result);
        }
        return result;
    }

private:
    void twoSum(int *l, int *r, int target, vector<vector<int>> &result) {
        while (l < r) {
            auto sum = (long) *l + *r;
            if (sum < target) l++;
            else if (sum > target) r--;
            else {
                result.push_back({-target, *l, *r});
                l++;
                r--;
                while (l < r && *l == *(l-1)) l++;
            }
        }
    }
};
// @lc code=end

